_posts/2023-09-15-kafka-producer-client-essentials.md

```diff
@@ -0,0 +1,323 @@
+For example, `localhost:9092` might be your starting point.
+In production, you can use a single load balancer, or a list of brokers.
+
+**TOPIC_NAME**
+
+The name of the topic where the producer client sends its messages.
+
+**NUM_MESSAGES**
+      
+This constant sets the number of messages the client produces before it shuts down.
+
+**MESSAGE_SIZE_BYTES**
+
+We'll use this constant to set the size of each message in bytes.
+
+**PROCESSING_DELAY_MS**
+
+Sometimes, it's good to slow things down a bit. 
+We can use this constant to add a delay in milliseconds between sending messages. 
+Adding a  delay can be useful when testing in order to simulate typical message processing time.
+
+These constants give us some control over the producer client's behavior. 
+
+### Example producer client
+
+Time to create our client.
+We want our example client to operate as follows:
+
+* Create a Kafka producer instance, which is responsible for sending messages to a Kafka topic.
+* Generate random message payloads, represented as byte arrays, that serve as the content of the messages being sent to Kafka cluster.
+* Use serializer classes that handle the transformation of message keys and values into a format suitable for the Kafka brokers. 
+* Control the rate at which messages are produced by introducing delays between each message using our `PROCESSING_DELAY_MS` value.
+* Handle errors that may occur during message transmission to the Kafka broker, determining when a message should be retried and when an error is considered non-recoverable. 
+
+**Producer configuration**
+
+We'll specify the minimal configuration properties required for a producer instance:
+
+* `BOOTSTRAP_SERVERS_CONFIG` for connection to the Kafka brokers. This picks up the value of the `BOOTSTRAP_SERVERS` constant.
+* `CLIENT_ID_CONFIG` that uses a randomly generated UUID as a client ID for tracking the source of requests.
+* `KEY_SERIALIZER_CLASS_CONFIG` and `VALUE_SERIALIZER_CLASS_CONFIG` to specify serializers that transform messages into a format suitable for Kafka brokers. 
+In this case, we'll specify the `ByteArraySerializer` as we want to transform byte array values.
+
+We'll also include methods that help with these operations:
+
+**`sleep` method**
+
+- Introduces a delay in the message-sending process for a specified number of milliseconds.
+- Useful for controlling the message production rate or simulating message processing times.
+- Handles potential `InterruptedException` if the thread responsible for sending messages is interrupted while paused.
+
+**`randomBytes` method**
+
+- Generates a random byte array of a specified size to serve as the payload for each message sent to the Kafka topic. Adds 65 to represent an uppercase letter in ASCII code (65 is 'A', 66 is 'B', and so on).
+- Ensures that the payload size is greater than zero and throws an `IllegalArgumentException` if not.
+
+**`retriable` method**
```

Comment:
Should we explain the impact of retrying? Can this cause duplicates? How does that differ from the producer configurations like `retries` or idempotency?

Reply:
The Kafka producer automatically handles retries for certain errors (i.e. connection errors), and you can tune it with configuration. That said, the application developer also needs to implement some retry logic for business level exceptions.

Yes, Kafka has at-least-once semantics by default, so the only way to avoid the risk of duplicates is to enable transactions. You can link our blog post for more details: https://strimzi.io/blog/2023/05/03/kafka-transactions

Reply:
Thanks @fvaleri -- I changed the note as follows:

> By default, Kafka operates with at-least-once delivery semantics, which means that messages can be delivered more than once in certain scenarios, potentially leading to duplicates. 
> To mitigate this risk, consider enabling transactions in your Kafka producer. 
> Transactions provide stronger guarantees of exactly-once delivery. 
> Additionally, you can use the `retries` configuration property to control how many times the producer will retry sending a message before giving up. 
> This setting affects how many times the `retriable` method may return true during a message send error. 
> For more information, see the [Strimzi post on transactions](https://strimzi.io/blog/2023/05/03/kafka-transactions/) and the [Strimzi documentation on ordered delivery](https://strimzi.io/docs/operators/latest/deploying#ordered_delivery).

Reply:
Fine, but I would remove the phrase about `retriable` method, which is unrelated (it comes after all retries are exhausted or timeout).